README.md
# Form Scoring Utility Instructions  
## Step 1: Hardware Setup
#### 1a - Receiver Setup. 
Plug microbit receiver into USB port on computer.
#### 1b - Upper Arm Accessory. 
Attach upper arm mount to arm  such that the microbit rests near the center of the deltoid. The exact position is not critical, but a greater distance between the upper arm microbit and the hand mircobit is preferred. When satisfied with the position of the upper arm mount, fasten the hook and loop closure such that the band does not slip, but circulation is not obstructed. Articulate your arm to ensure the band is not restrictive. If there is any restirction, loosen the hook and loop closure. Do not exercise with an armband that is too tight.
#### 1c - Glove Accessory. 
Put on glove, and articulate the hand and wrist to ensure a full range of motion is possible without restriction. If the battery pack limits the motion of the hand, reposition the battery pack to a different location on the cuff of the glove, and perform articulation test again. Repeat position testing until a comfortable and non-restrictive position for the battery pack is found.
## Step 2: Starting Wekinator  
#### Open WekinatorProject.wekproj in Wekinator 
(WekinatorProject.wekproj can be found in FinalProject-FormScoring/WekinatorProject)  
## Step 3: Starting gui.py
#### Run gui.py
Using Unix Terminal:  
~/FinalProject-FormScoring$ python gui.py
## Step 4: Using the GUI
![Image of GUI](https://github.com/CUBoulder-2019Sp-IML4HCI/FinalProject-FormScoring/blob/master/img/gui_instructions.png)
### Training the Model
1. Using the dropdown (displayed above with 'Still'), select the form/failure you wish to train.
2. Using the demo animation as a guide, record a single rep of the form/failure by clicking 'Start/Stop Recording' at the beginning of the rep, and again when the rep is complete.
3. Record at least 2 samples for each form/failure.
4. Repeat until all form/failure modes have been trained.

### Using the Model
To use the model, click 'Start/Stop Running' when beginning a set, and again when you complete your set.

### Reading the Results
The frequency of good reps and failures will be displayed at the bottom of the GUI. 


changelog.md
## Changelog  
# 03-06-2019 Ian Wilkins: 
- Finish project proposal draft
                        
# 03-07-2019 Ian Wilkins: 
- Begin prototyping microbit chain.
- Create initial mircobit reciever (receive -> write to serial) hex
- Create initial microbit middle link (receive -> package -> send) hex
- Create initial microbit end link (send via radio) hex
- Create initial python OSC server
                                          
# 03-12-2019 Ian Wilkins: 
- Update microbit middle link (receive -> package -> send) hex
                      
# 03-13-2019 Ian Wilkins: 
- Add link to GitHub repo to project proposal
- Add acceptance criteria to project milestones in project proposal
- Make final revisions to project proposal and submit to Moodle
                        
# 03-14-2019 Ian Wilkins: 
- After meeting with Ben, I have been shown that a microbit chain is not going to be as feasible as 2 microbits communicating independently with the receiver.
- Begin prototyping signal strength reading
- Begin new python OSC server to handle new message format.
                        
# 03-14-2019 Garrett Sippel:
- Wrote ocs handler

# 03-17-2019 Ian Wilkins: 
- Create first working microbit 'a' and 'b' hex
- Update python OSC server to interpret messages from microbit 'a' and 'b'
                        
# 03-20-2019 Both:        
- Created youtube video
- Link: https://youtu.be/LxvbDUXSePE

# 03-25-2019 Ian Wilkins: 
- Create first glove and armband mount for 'a' and 'b' microbits. The following discoveries were made from these prototypes:
  1. Thick/robust work gloves are more rigid (and therein easier to mount a microbit to), but limit                                dexterity during exercises, making them less ideal.
  2. A loose fitting glove feels less awkward when a microbit is mounted, but poses potential safety                              issues when dealing with heavy weights (maximum grip strength is reduced).
  3. It is VERY EASY to make an arm band that cuts off circulation. The current battey pack weighs                                enough that supporting it is problematic. Potential solutions include decreasing tension of the                              cuff, but increasing the total contact surface area.


# 03-27-2019 Ian Wilkins: 
- Create second glove and armband mount for 'a' and 'b' microbits. The following discoveries were made from these prototypes:
    1. Thin and stretchy gloes have less restriction to dexterity, but are heard to mount a microbit                                to. The strech factor complicates using static mounting points. Too much stiching will snap when                              the glove is stretched, but too little stiching will not adequately constrain the microbit in 3D                              space.
    2. A tight fitting glove holds the microbit mounting method closer to the skin. In order to reduce                              irritation, fine thread was used to adhere the microbit.
    3. The battery pack is bulky, and position preference seems to change with each user. I need to                                find a way to have a repositional battery pack that is still safely secured. 
    4. The next iteration of the upper arm cuff seemed promising until it was taken to the gym. An                                  unforseen problem was realized: the microbit needs to be able to be detached so the band can be                              cleaned (the band band will absorb sweat during a workout. Current material and design is not                                well suited for easy washing.

# 03-31-2019 Garrett Sippel:
- Create initial GUI.

# 04-01-2019 Garrett Sippel:
- Experiment with threading GUI and server class together.


# 04-01-2019 Ian Wilkins:
- At this point in the project, we have reasonable mounts for the microbits (though more development is needed on the upper arm mount), we have a basic GUI, we have basic data handling, and we have all the components talking to each other. We have confirmed that we can recieve and process the XYZ acceleration values and signal stength between microbits, and that the signal strength is in fact reasonable different at different stages of our test exercises. The next step is to implement the DTW. Today involved reasearch on available DTW libraries and what option is most applicable to our project.

# 04-02-2019 Ian Wilkins: 
- Create third glove and armband mount for 'a' and 'b' microbits. The following discoveries were made from these prototypes:
    1. Constraining the microbits with the auxillary pins parallel to the knuckles on the ulnar side of                              the glove, and a slightly slackened thread through the power inpit connecting to the space                                    between the tendon of the thumb and the tendon of the forefinger provides adequate constraint,                                handles stretching well, and does not feel awkward during use.
    2. A glove was found with a cuff material that adheres well to the hook side of a hook and loop                                  patch. The battery pack now has a hook patch added so it can bea easily repositioned anywhere on                              the cuff of the glove without adding extra bulk.
    3. The glove is almost done, it now needs a way for the microbit to be detached when the glove is                                whashed.

ocsfinalproject.py
# requires the following python libraries:
#  python-osc
#  pyserial

import os
import serial
from pythonosc import udp_client

class Sender:
    def __init__(self):
        self.output = [None]*8
        self.send = self.wait_for_fill
    def wait_for_fill(self):
        if None in self.output:
            print("Waiting on full output.")
        else:
            self.send = self.send_signal
    def send_signal(self):
        print(self.output)
        client.send_message("/final", self.output)


# change this string depending upon where your computer makes a device for the micro:bit
if os.name == 'nt':
    serialport = "COM4"
else:
    serialport = "/dev/cu.usbmodem14302"

 
ser = serial.Serial(serialport, 115200)
client = udp_client.SimpleUDPClient("localhost", 8999)
S = Sender()
while(True):
    line = ser.readline().decode('utf8').strip()

    message = list(map(float, line.split("^")))
    tag = message[0]
    if tag == "a":
        S.output[0:3] = message[1:3]
    elif tag == "b":
        S.output[4:7] = message[1:3]
    S.send()
    print(S.output)


#import serial
#from pythonosc import udp_client

# change this string depending upon where your computer makes a device for the micro:bit
#serialport = "/dev/ttyACM0"

#ser = serial.Serial(serialport, 115200)
#client = udp_client.SimpleUDPClient("localhost", 8999)

#while(True):
#    try:
#        line = ser.readline().decode('utf8').strip()
#        data = list(map(float, line.split("^")))
#    except:
#        data = []
#        
#    if(len(data) == 6):        
#        print(f"({data[0]}, {data[1]}, {data[2]}, {data[3]}, {data[4]}, {data[5]})")

oscfinalproject.py
# requires the following python libraries:
#  python-osc
#  pyserial

import os
import serial
from pythonosc import udp_client

class Sender:
    def __init__(self):
        self.output = [None]*8
        self.send = self.wait_for_fill
    def wait_for_fill(self):
        if None in self.output:
            print("Waiting on full output.")
            #print(self.output)
        else:
            self.send = self.send_signal
    def send_signal(self):
        print(self.output)
        client.send_message("/wek/inputs", self.output)


#if windows
if os.name == 'nt':
    serialport = "COM7"
#else linux
else:
    serialport = "/dev/cu.usbmodem14302"

 
ser = serial.Serial(serialport, 115200)
client = udp_client.SimpleUDPClient("localhost", 6448)
S = Sender()
while(True):
    line = ser.readline().decode('utf8').strip()

    message = list(map(str, line.split("^")))
    tag = message[0]
    print(message)
    if tag == "a":
        S.output[:4] = message[1:]
    elif tag == "b":
        S.output[4:] = message[1:]
    S.send()
    #print(S.output)


#import serial
#from pythonosc import udp_client

# change this string depending upon where your computer makes a device for the micro:bit
#serialport = "/dev/ttyACM0"

#ser = serial.Serial(serialport, 115200)
#client = udp_client.SimpleUDPClient("localhost", 8999)

#while(True):
#    try:
#        line = ser.readline().decode('utf8').strip()
#        data = list(map(float, line.split("^")))
#    except:
#        data = []
#        
#    if(len(data) == 6):        
#        print(f"({data[0]}, {data[1]}, {data[2]}, {data[3]}, {data[4]}, {data[5]})")

formScoreBits_to_osc.py
# requires the following python libraries:
#  python-osc
#  pyserial
 
import serial
from pythonosc import udp_client
 
# change this string depending upon where your computer makes a device for the micro:bit
serialport = "/dev/ttyACM0"
 
ser = serial.Serial(serialport, 115200)
client = udp_client.SimpleUDPClient("localhost", 8999)
 
while(True):
    line = ser.readline().decode('utf8').strip()
    print(line)
    
    if line.count('^') != 4: continue
 
    #n, s, x, y, z = map(str, line.split("^"))
    n, s, x, y, z = line.split("^")
 
    print(f"({n}, {s}, {x}, {y}, {z})")
 
    client.send_message("/form_data", [n, s, x, y, z] )

#import serial
#from pythonosc import udp_client

# change this string depending upon where your computer makes a device for the micro:bit
#serialport = "/dev/ttyACM0"

#ser = serial.Serial(serialport, 115200)
#client = udp_client.SimpleUDPClient("localhost", 8999)

#while(True):
#    try:
#        line = ser.readline().decode('utf8').strip()
#        data = list(map(float, line.split("^")))
#    except:
#        data = []
#        
#    if(len(data) == 6):        
#        print(f"({data[0]}, {data[1]}, {data[2]}, {data[3]}, {data[4]}, {data[5]})")
#        client.send_message("/wand_accel", data )


gui.py
from pythonosc import dispatcher
from pythonosc import osc_server
import threading
from tkinter import *
import serial
from pythonosc import udp_client
import os
import time


class GuiLogic:
    def __init__(self):
        self.master = Tk()
        self.exercises = ["Still","Bicep Curl-Good","Bicep Curl-Bad(pronation)","Bicep Curl-Bad(half)"]
        self.exercise_gifs = ["placeholder","img/still.gif","img/BicepCurl_GoodForm_1Arm.gif","img/BicepCurl_PronationFailure_crop.gif","img/BicepCurl_HalfRep_1Arm.gif"]
        self.exercise_gifs_length = [0,1,41,3,17]
        self.exercise_speed = [0,100,100,300,100]
        self.exercise_values = {e:i+1 for i,e in enumerate(self.exercises)}
        self.selected_exercise = StringVar(self.master)
        self.selected_exercise.set(self.exercises[0])
        exercise_select = OptionMenu(self.master, self.selected_exercise,*self.exercises)
        record_button = Button(self.master, text="Start/Stop Recording",command=self.record)
        run_button = Button(self.master, text="Start/Stop Running",command = self.run)
        exercise_select.pack()
        record_button.pack()
        run_button.pack()
        self.isRunning = BooleanVar(self.master)
        self.isRunning.set(False)
        self.recording = BooleanVar(self.master)
        self.recording.set(False)
        self.loop = True
        self.e = threading.Event()
        self.data = []
        self.gifs_dict = {i+1:self.load_gif_by_value(i+1) for i,e in enumerate(self.exercises)}
        self.frames = self.gifs_dict[self.getExerciseValue()]
        self.label = Label(self.master)
        self.label.pack()
        #self.rep_text = Label(self.master, text = "Number of Reps" )
        #self.rep_text.pack()
        #self.reps_entry = Entry(self.master)
        #self.reps_entry.insert(END, '5')
        #self.reps_entry.pack()
        self.history = []

        self.good_counts = IntVar(self.master,0)
        self.half_counts = IntVar(self.master,0)
        self.pronation_counts = IntVar(self.master,0)

        self.good_rep_mesg = StringVar(self.master, value= "Good Reps Counted: %i" % self.good_counts.get())
        self.half_rep_mesg = StringVar(self.master, value="Half Reps Counted: %i" % self.half_counts.get())
        self.pronation_rep_mesg = StringVar(self.master, value="Failures to Pronate Counted: %i" % self.pronation_counts.get())

        self.good_rep_disp = Label(self.master,textvariable = self.good_rep_mesg)
        self.half_rep_disp = Label(self.master, textvariable= self.half_rep_mesg)
        self.pronation_rep_disp = Label(self.master, textvariable=self.pronation_rep_mesg)

        self.good_rep_disp.pack()
        self.half_rep_disp.pack()
        self.pronation_rep_disp.pack()
    def load_gif_by_value(self,value):
        file_name = self.exercise_gifs[value]
        return [PhotoImage(master=self.master, file=file_name, format='gif -index %i' % i) for i in range(self.exercise_gifs_length[value])]

    def run(self):
        if self.isRunning.get():
            self.isRunning.set(False)
            self.analyze_set()
        else:
            self.isRunning.set(True)

    def record(self):
        if self.recording.get():
            self.stop_record()
        else:
            self.start_record()

    def start_record(self):
        print("started recording")
        self.recording.set(True)

    def stop_record(self):
        print("stopped recording")
        self.recording.set(False)

    def getExerciseValue(self):
        v = self.exercise_values[self.selected_exercise.get()]
        return v

    def update_gif(self, ind, prev_image):
        if prev_image != self.getExerciseValue():
            prev_image = self.getExerciseValue()
            ind = 0
            self.frames = self.gifs_dict[self.getExerciseValue()]
        frame = self.frames[ind]
        ind = (ind + 1)%self.exercise_gifs_length[prev_image]
        self.label.configure(image=frame)
        self.master.after(self.exercise_speed[prev_image], self.update_gif, ind, prev_image)

    def receive_data(self,head,one,two,three,four):
        print("received this data: ", head,one,two,three,four)

    def receive_class(self,head):
        if head == "/output_1":
            self.history.append((1,time.time()))
        elif head == "/output_2":
            self.history.append((2, time.time()))
        elif head == "/output_3":
            self.history.append((3, time.time()))
        else:
            self.history.append((4, time.time()))
        print(self.history)

    def analyze_set(self):
        delay = 2
        self.good_counts.set(0)
        self.half_counts.set(0)
        self.pronation_counts.set(0)
        good_reps_counted = 0
        half_reps_counted = 0
        pronation_reps_counted = 0
        last_good = 0
        good = [rep[1] for rep in self.history if rep[0] == 2]
        half = [rep[1] for rep in self.history if rep[0] == 4]
        true_half = []
        pro = [rep[1] for rep in self.history if rep[0] == 3]
        true_pro = []
        for h in half:
            flag = False
            for g in good:
                if abs(h - g) <= delay:
                    flag = True
            if not flag:
                true_half.append(h)
        for p in pro:
            flag = False
            for g in good:
                if abs(p-g) <= delay:
                    flag = True
            if not flag:
                true_pro.append(p)

        self.good_counts.set(len(good))
        self.half_counts.set(len(true_half))
        self.pronation_counts.set(len(true_pro))

        self.good_rep_mesg.set("Good Reps Counted: %i" % self.good_counts.get())
        self.half_rep_mesg.set("Half Reps Counted: %i" % self.half_counts.get())
        self.pronation_rep_mesg.set("Failures to Pronate Counted: %i" % self.pronation_counts.get())

        self.history = []


    """
    def analyze_set(self):
        delay = 2
        self.good_counts.set(0)
        self.half_counts.set(0)
        self.pronation_counts.set(0)
        good_reps_counted = 0
        half_reps_counted = 0
        pronation_reps_counted = 0
        last_good = 0

        for i in range(len(self.history)):
            if self.history[i][0] == 2:
                good_reps_counted += 1
                last_good = self.history[i][1]
            elif self.history[i][0] == 3:
                start = i-2
                if start<0:
                    start = 0
                flag = False
                for j in range(start,len(self.history)):
                    if self.history[j][0] == 2:
                        t = abs(self.history[i][1] - self.history[j][1])
                        print(t)
                        if t > delay:
                            pronation_reps_counted += 1
                            flag = True
                            break
                if flag == False:
                    pronation_reps_counted += 1
                #t = abs(last_good - self.history[i][1])
                #print(t)
                #if t > 4:
                #   pronation_reps_counted += 1
            elif self.history[i][0] == 4:
                start = i-2
                if start<0:
                    start = 0
                flag = False
                for j in range(start,len(self.history)):
                    if self.history[j][0] == 2:
                        t = abs(self.history[i][1] - self.history[j][1])
                        print(t)
                        if t > delay:
                            half_reps_counted += 1
                            flag = True
                            break
                if flag == False:
                    half_reps_counted += 1
                #t = abs(last_good - self.history[i][1])
                #print(t)
                #if t > 4:
                #   half_reps_counted += 1
                

        #half_reps_counted = half_reps_counted - good_reps_counted
        #if half_reps_counted < 0:
        #    half_reps_counted = 0
        """








def str_float_map(x):
    try:
        y = float(x)
    except:
        y = str(x)
        
    return(y)


class Server:
    def __init__(self, GL, ip="localhost", port=6448):
        self.GL = GL
        self.ip = ip
        self.port = port
        self.client = udp_client.SimpleUDPClient(self.ip, self.port)
        self.output = [None] * 8
        self.send = self.wait_for_fill
        self.record_prev_state = False
        self.run_prev_state = False

    def wait_for_fill(self):
        if None in self.output:
            print("Waiting on full output.")
        else:
            self.send = self.send_signal

    def send_signal(self):
        s1 = self.output[0]
        s2 = self.output[4]
        if s1 == 0 or s2 == 0:
            strength = max(s1,s2)
        else:
            strength = (s1+s2)/2
        o = [strength] + self.output[1:4] + self.output[5:]
        #print(o)
        self.client.send_message("/wek/inputs", o)

    def run_server(self):
        # if windows
        if os.name == 'nt':
            serialport = "COM7"
        # else linux
        else:
            #serialport = "/dev/cu.usbmodem14302"
            serialport = "/dev/ttyACM0"

        ser = serial.Serial(serialport, 115200)
        while (True):
            line = ser.readline().decode('utf8').strip()
            #message = list(map(str, line.split("^")))
            message = list(map(str_float_map, line.split("^")))
            
            tag = message[0]
            #print(message)
            if tag == "a":
                self.output[:4] = message[1:]
            elif tag == "b":
                self.output[4:] = message[1:]
            self.send()
            new_rec_state = GL.recording.get()
            new_run_state = GL.isRunning.get()

            if new_run_state != self.run_prev_state:
                self.run_prev_state = new_run_state
                if new_run_state:
                    self.client.send_message("/wekinator/control/startRunning", 1)
                else:
                    self.client.send_message("/wekinator/control/stopRunning", 1)

            if new_rec_state != self.record_prev_state:
                self.record_prev_state = new_rec_state
                if new_rec_state:
                    self.client.send_message("/wekinator/control/startDtwRecording", GL.getExerciseValue())
                else:
                    self.client.send_message("/wekinator/control/stopDtwRecording", 1)


class Listener:
    def __init__(self,GL):
        self.GL = GL
    def run_listener(self):
        d = dispatcher.Dispatcher()
        #d.map("/wek/outputs", self.GL.receive_data)
        d.map("/output_1", self.GL.receive_class)
        d.map("/output_2", self.GL.receive_class)
        d.map("/output_3", self.GL.receive_class)
        d.map("/output_4", self.GL.receive_class)
        port = 12000
        ip = "127.0.0.1"
        server = osc_server.ThreadingOSCUDPServer((ip, port), d)
        print("Serving on {}".format(server.server_address))
        server.serve_forever()


GL = GuiLogic()
S = Server(GL)
L = Listener(GL)
Server_Thread_L = threading.Thread(target = L.run_listener , args=())
Server_Thread_L.daemon = True
Server_Thread_L.start()
Server_Thread_S = threading.Thread(target = S.run_server , args=())
Server_Thread_S.daemon = True
Server_Thread_S.start()
GL.master.after(0, GL.update_gif, 0, GL.getExerciseValue())
GL.master.mainloop()